practice_project/weather_forecast.py

```diff
@@ -0,0 +1,86 @@
+import requests
+from abc import ABC, abstractmethod
+
+class WeatherDataProvider(ABC):
+    @abstractmethod
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Abstract method to fetch weather forecast for a given latitude and longitude.
+        """
+        pass
+      
+class WeatherAPIDataProvider(WeatherDataProvider):
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Fetch weather forecast using Open-Meteo API.
+        """
+        api_url = f"https://api.open-meteo.com/v1/forecast?latitude={latitude}&longitude={longitude}&hourly=temperature_2m"
+        response = requests.get(api_url)
+        if response.status_code == 200:
+            return response.json()
+        else:
+            raise Exception("Failed to fetch weather forecast from API")
+          
+class MockWeatherDataProvider(WeatherDataProvider):
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Mocked method to return a dummy weather forecast.
+        """
+        return {
+            "latitude": latitude,
+            "longitude": longitude,
+            "hourly": {}
+        }
+        
+class WeatherDataHandler:
+    def __init__(self, data_provider):
+        """
+        Initialize WeatherDataHandler with a data provider.
+        """
+        self.data_provider = data_provider
+
+    def get_weather_forecast(self, latitude, longitude):
+        """
+        Get weather forecast for a given latitude and longitude.
+        """
+        return self.data_provider.fetch_weather_forecast(latitude, longitude)
+      
+class FancyWeatherDataHandler(WeatherDataHandler):
+    def __init__(self, data_provider):
+        """
+        Initialize FancyWeatherDataHandler with a data provider.
+        """
+        super().__init__(data_provider)
+
+    def fancy_get_weather_forecast(self, latitude, longitude):
+        """
+        A fancy way to get weather forecast.
+        """
+        print("Fetching weather forecast in a fancy way...")
+        return self.get_weather_forecast(latitude, longitude)
```

Comment:
i like how you implement different handlers, but you should most likely re-write this so that you have an interface class declaring an abstract method `handle_weather_data(self, weather_data)`.

It doesn't make too much sense here to extend `WeatherDataHandler` in your `FancyWeatherDataHandler` since you don't really re-use anything